src/infra/core/src/pull_service_impl.rs

```diff
@@ -419,20 +422,64 @@ impl PullServiceImpl {
         Ok(results)
     }
 
+    #[async_recursion::async_recursion]
     async fn transform_multi(
         &self,
         batch: &[PullItem], // TODO: Move to avoid cloning
-        listener: Option<Arc<dyn TransformMultiListener>>,
+        transform_listener: Option<Arc<dyn TransformMultiListener>>,
+        compacting_listener: Option<Arc<dyn CompactingMultiListener>>,
+        retry_with_compaction: bool,
     ) -> Result<Vec<PullResponse>, InternalError> {
         let transform_requests = batch.iter().map(|pi| pi.local_ref.clone()).collect();
 
         let transform_results = self
             .transform_svc
-            .transform_multi(transform_requests, listener)
+            .transform_multi(transform_requests, transform_listener.clone())
             .await;
 
         assert_eq!(batch.len(), transform_results.len());
 
+        if retry_with_compaction {
+            let mut compacting_target_refs = vec![];
+            for (dataset_ref, transform_result) in &transform_results {
+                if let Err(transform_err) = transform_result
+                    && let TransformError::InvalidInterval(_) = transform_err
+                {
+                    compacting_target_refs.push(dataset_ref.clone());
+                }
+            }
+
+            if !compacting_target_refs.is_empty() {
+                let compacting_results = self
+                    .compacting_svc
+                    .compact_multi(
+                        compacting_target_refs,
+                        CompactingOptions {
+                            keep_metadata_only: true,
+                            ..Default::default()
+                        },
+                        compacting_listener,
+                    )
+                    .await;
+
+                let retry_items: Vec<_> = std::iter::zip(batch, compacting_results)
+                    .filter_map(|(pi, response)| {
+                        assert_eq!(pi.local_ref, response.dataset_ref);
+                        if let Ok(result) = response.result
+                            && let CompactingResult::Success { .. } = result
+                        {
+                            return Some(pi.clone());
+                        }
+                        None
+                    })
+                    .collect();
+
+                return self
+                    .transform_multi(&retry_items, transform_listener, None, false)
+                    .await;
```

Comment:
Similarly here, we need to return the combined result of datasets that succeeded the transform the first time and those that were reset and then transformed. Currently you will return only results for those that were reset.

The logic of combining results may get too complex... As an option - I'd consider making `reset_on_input_divergence` a parameter of the `TransformService` (as an options struct passed into `transform_multi`). This way it can handle divergence internally, and `PullService` will not need to deal with complexity of resetting, retrying, and merging results - retries can happen on per-dataset basis inside `TrnasformServiceImpl`.

What do you think @zaychenko-sergei ?

Reply:
I would have completely placed the entire algorithm of "error-compact-retry" into `TransformServiceImpl`. Pull is at a higher level of abstraction.